src/lib.rs

```diff
@@ -1,21 +1,392 @@
+                        return Ok(());
+                    }
+                }
+
+                StarlingMessage::TaskErrored(id, error) => {
+                    assert!(self.tasks.remove(&id).is_some());
+
+                    let thread_id = id.into();
+                    let join_handle = self.threads
+                        .remove(&thread_id)
+                        .expect("should have a thread join handle for the errored task");
+                    join_handle
+                        .join()
+                        .expect("should join errored task's thread OK");
+
+                    if id == main.id() {
+                        // TODO: notification of shutdown and joining other threads and things.
+                        return Err(error);
+                    }
+                }
+
+                StarlingMessage::NewTask(_task_handle, _join_handle) => unimplemented!(),
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Messages that threads can send to the starling main thread.
+#[derive(Debug)]
+pub(crate) enum StarlingMessage {
+    /// The task on the given thread completed successfully.
+    TaskFinished(task::TaskId),
+
+    /// The task on the given thread failed with the given error.
+    TaskErrored(task::TaskId, Error),
+
+    /// A new child task was created.
+    NewTask(task::TaskId, thread::JoinHandle<()>),
+}
+
+/// A handle to the Starling system.
+///
+/// A `StarlingHandle` is a capability to schedule IO on the event loop, spawn
+/// work in one of the utility thread pools, and communicate with the Starling
+/// main thread. Handles can be cloned and sent across threads, propagating
+/// these capabilities.
+#[derive(Clone)]
+pub(crate) struct StarlingHandle {
+    options: Arc<Options>,
+    sync_io_pool: CpuPool,
+    cpu_pool: CpuPool,
+    sender: Sender<StarlingMessage>,
+}
+
+impl fmt::Debug for StarlingHandle {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "StarlingHandle {{ .. }}")
+    }
+}
+
+impl StarlingHandle {
+    /// Get the `Options` that this starling system was configured with.
+    pub fn options(&self) -> &Arc<Options> {
+        &self.options
+    }
+
+    /// Get a handle to the thread pool for shimming synchronous IO (perhaps
+    /// from a library that wasn't written to be async) into the system.
+    pub fn sync_io_pool(&self) -> &CpuPool {
+        &self.sync_io_pool
+    }
+
+    /// Get a handle to the thread pool for performing CPU bound background
+    /// tasks.
+    pub fn cpu_pool(&self) -> &CpuPool {
+        &self.cpu_pool
+    }
+
+    /// Send a message to the starling main thread.
+    pub fn send(
+        &self,
+        msg: StarlingMessage,
+    ) -> futures::sink::Send<futures_mpsc::Sender<StarlingMessage>> {
+        self.sender.clone().send(msg)
+    }
+
+    /// Send a message to the starling main thread synchronously.
+    pub fn send_sync(&self, msg: StarlingMessage) -> Result<()> {
```

Comment:
Do we need the ability to send sync messages? I'd at least try to get away without it as long as possible. The error message below looks like it could be sent asynchronously?